archipelago/pipeline.py

```diff
@@ -0,0 +1,1028 @@
+            netlist_read = True
+        elif "ID to Names:" in line:
+            netlist_read = False
+            id_to_name_read = True
+        elif "Netlist Bus:" in line:
+            netlist_read = False
+            id_to_name_read = False
+        elif netlist_read:
+            if len(line.split(":")) > 1:
+                edge_id = line.split(":")[0]
+                connections = line.split(":")[1]
+
+                connections = re.findall(r'\b\S+\b', connections)
+
+                netlist[edge_id] = []
+                for conn1, conn2 in zip(connections[::2], connections[1::2]):
+                    netlist[edge_id].append((conn1, conn2))
+        elif id_to_name_read:
+            if len(line.split(":")) > 1:
+                id = line.split(":")[0]
+                name = line.split(":")[1]
+                id_to_name[id] = name.strip()
+              
+    return netlist, id_to_name
+
+def load_folded_regs(folded_file):
+    f = open(folded_file, "r")
+    lines = f.readlines()
+    pe_reg = set()
+ 
+    for line in lines:
+        reg_entry = re.findall(r'\b\S+\b', line.split(":")[0])
+        entry = re.findall(r'\b\S+\b', line.split(":")[1])
+        blk_id = entry[0]
+        port = entry[-1]
+        if reg_entry[0][0] == 'r' and blk_id[0] == 'p':
+            pe_reg.add(((reg_entry[0], reg_entry[1]),(blk_id, port)))
+
+    return pe_reg
+
+def load_shift_regs(shift_regs_file, pe_reg):
+    shift_regs = set()
+    folded_regs = {reg:pe for (reg,_),pe in pe_reg}
+
+    f = open(shift_regs_file, "r")
+    lines = f.readlines()
+    pe_reg = set()
+ 
+    for line in lines:
+        id = line.strip()
+        shift_regs.add((id, None))
+        if id in folded_regs:
+            shift_regs.add(folded_regs[id])        
+    return shift_regs
+
+def segment_to_node(segment, net_id):
+    if segment[0] == "SB":
+        track, x, y, side, io_, bit_width = segment[1:]
+        node1 = pnr_graph.Node("route", x, y, route_type = "SB", track = track, side = side, io = io_, bit_width = bit_width, net_id = net_id)
+    elif segment[0] == "PORT":
+        port_name, x, y, bit_width = segment[1:]
+        node1 = pnr_graph.Node("route", x, y, route_type = "PORT", bit_width = bit_width, net_id = net_id, port = port_name)
+    elif segment[0] == "REG":
+        reg_name, track, x, y, bit_width = segment[1:]
+        node1 = pnr_graph.Node("route", x, y, route_type = "REG", track = track, bit_width = bit_width, net_id = net_id, reg_name = reg_name)
+    elif segment[0] == "RMUX":
+        rmux_name, x, y, bit_width = segment[1:]
+        node1 = pnr_graph.Node("route", x, y, route_type = "RMUX", bit_width = bit_width, net_id = net_id, rmux_name = rmux_name)
+    else:
+        raise ValueError("Unrecognized route type")
+    return node1
+
+def get_tile_at(x, y, bw, placement, port = "", reg = False):
+    pond_ports = ["data_in_pond_0", "data_out_pond_0", "flush"]
+
+    for tile_id, place in placement.items():

```

Comment:
This doesn't look efficient. you have to search the entire placement for every x and y. Consider build a reverse-indexed placement first, e.g. (x, y) -> list[nodes]